### PR TITLE
Implement the two kinds of autocorrelation consumers + other small changes

### PR DESCRIPTION
**Pull request by MantautasRimkus**

### CODE REVIEW

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+#include <boost/numeric/ublas/matrix.hpp>
+
+
+namespace SampleFlow
+{
+  namespace Consumers
+  {
+    /**
+     * A Consumer class that implements computing the running sample autocovariance function:
+     * \hat\gamma(k)=\frac{1}{n}\sum_{t=1}^{n-k}{(x_{t+k}-\bar{x})(x_{t}-\bar{x})}
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: You can use latex formulas in the documentation if you enclose them by `$...$`.

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+     *
+     * ### Threading model ###
+     *
+     * The implementation of this class is thread-safe, i.e., its
+     * consume() member function can be called concurrently and from multiple
+     * threads.
+     *
+     *
+     * @tparam InputType The C++ type used for the samples $x_k$. In
+     *   order to compute covariances, the same kind of requirements
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]:
```suggestion
     *   order to compute auto-covariances, the same kind of requirements
```

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+     *   have to hold as listed for the Covariance class.
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: But actually, you're only considering the 1d case here, so the requirements on the data type may be different. You might want to think about what kind of type `InputType` can actually be.

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+     */
+    template <typename InputType>
+    class autocovariancedim1 : public Consumer<InputType>
+    {
+      public:
+        /**
+         * The data type of the elements of the input type.
+         */
+
+        using scalar_type = InputType;
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: The documentation immediately above corresponds to the `using` declaration here. Remove the empty line to make that clear, and make sure that there is an empty line after the declaration and the following comment.

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        /**
+         * The type of the information generated by this class, i.e., in which
+         * the autocovariance function is computed. By itself, for autocovariance there should
+         * be enough to use vector. However, because we keep updating autocovariance, we need to save
+         * parts used in calculation
+         */
+
+        using value_type = boost::numeric::ublas::matrix<scalar_type>;
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: Same comment about spacing here.

Since you're in 1d, isn't the auto-covariance just a vector where the elements are indexed by `k`? In that case, some kind of vector data type seems more appropriate, e.g., the `std::vector` class.

File: include/sampleflow/consumers/autocovariance_dim_1.h (hunk 4 above)
Comment (bangerth) [reply include/sampleflow/consumers/autocovariance_dim_1.h]: On a separate note, your class can only be used if `InputType` is a scalar type (not a vector). You can verify that this is actually so by adding the following two lines (stolen from the `Histogram` class declaration in `consumers/histogram.h`):
```
       static_assert (std::is_arithmetic<InputType>::value == true,
                       "This class can only be used for scalar input types.");
```

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+        /**
+         * Constructor.
+         */
+
+        autocovariancedim1();
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: same about spacing here and probably in other places below as well.

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+
+        /**
+         * Process one sample by updating the previously computed covariance
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: might want to go through the documentation and find all places where it talks about "covariance" and update that as appropriate.

File: include/sampleflow/consumers/autocovariance_dim_1.h (hunk 6 above)
Comment (bangerth) [reply include/sampleflow/consumers/autocovariance_dim_1.h]: I think the constructor should probably get an argument that denotes the length of the auto-covariance vector. You don't want to compute an infinitely long auto-covariance, but one of fixed length that the user can choose.

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+         * of this class.
+         */
+
+        InputType           current_mean;
+
+        /**
+         * Currently describes lag of autocovariance
+         */
+
+        InputType 			k;
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: I suspect that you probably want to move this variable to a different place, but in any case, the data type needs to be an integer type, not whatever the type of the samples is. I would choose `unsigned int`, because you describe a lag or length, and that's a positive integer.

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+
+        /**
+         * The current value of $\bar x_k$ as described in the introduction
+         * of this class.
+         */
+        boost::numeric::ublas::matrix<scalar_type> current_autocovariationdim1_matrix;
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: Isn't `\bar x` the mean value? In that case, it ought to have data type `InputType`. Or maybe I misunderstand.

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+
+        /**
+        * Save previous sample value
+        */
+
+        boost::numeric::ublas::matrix<scalar_type> past_sample;
+        boost::numeric::ublas::matrix<scalar_type> past_sample_replace;
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: same here

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+        {
+          n_samples = 1;
+          current_autocovariationdim1_matrix.resize (k,3);
+          for (unsigned int i=0; i<k; ++i)
+        	for (unsigned int j=0; j<2; ++j){
+        		current_autocovariationdim1_matrix(i,j) = 0;
+        	}
+          current_mean = sample;
+          past_sample.resize (k,1);
+          past_sample_replace.resize (k,1);
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: Ah, now I see how you are doing this -- you are creating a `k x 1` matrix. This is probably easiest handled with a `std::vector` instead.

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+          past_sample(0,0) = sample;
+        }
+      else if(n_samples < k)
+		{
+     	 ++n_samples;
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: you ought to let eclipse indent the file for you so that things are nicely aligned :-)

File: include/sampleflow/consumers/autocovariance_dim_1.h [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+    	// Then also update the running mean:
+    	 InputType update = sample;
+    	          update -= current_mean;
+    	          update /= n_samples;
+    	          current_mean += update;
+		}
+     else
+     {
+        ++n_samples;
```
Comment (bangerth) [inline include/sampleflow/consumers/autocovariance_dim_1.h]: You get into this `else` branch if `n_samples>=k`. In that case, you probably just want to throw away the first sample you stored. I think you do that below in line 220, but there are easier ways. For example, using the `std::list` or `std::dequeue` data structures, there are `push_back()` and `pop_front()` functions that allow you to do exactly what you want to do. We can talk about this on Monday.

File: include/sampleflow/consumers/spurious_autocovariance_dim_n.h [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+  {
+    /**
+     * NOTICE: We call results as spurious autocovariance, because by definition it is not autocovariance.
+     * A Consumer class that implements computing the running sample spurious autocovariance function:
+     * \hat\gamma(l)=\frac{1}{n}\sum_{t=1}^{n-l}{(\bm{x}_{t+l}-\bar\bm{x})(bm{x}_{t}-\bar\bm{x})}
+     *
+     *This code for every new sample updates \hat\gamma(k), l=1,2,3...,k. Choice of k can be done
+     *by setting it below.
+     *%%%%%%%%%%%%Manty's comment: It would be great to set k in mcmc_test.cc, not here. I don't know
+     *how to set it globally.
```
Comment (bangerth) [inline include/sampleflow/consumers/spurious_autocovariance_dim_n.h]: using the constructor argument I mention above.

File: include/sampleflow/consumers/spurious_autocovariance_dim_n.h [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+
+
+namespace SampleFlow
+{
+  namespace Consumers
+  {
+    /**
+     * NOTICE: We call results as spurious autocovariance, because by definition it is not autocovariance.
+     * A Consumer class that implements computing the running sample spurious autocovariance function:
+     * \hat\gamma(l)=\frac{1}{n}\sum_{t=1}^{n-l}{(\bm{x}_{t+l}-\bar\bm{x})(bm{x}_{t}-\bar\bm{x})}
```
Comment (bangerth) [inline include/sampleflow/consumers/spurious_autocovariance_dim_n.h]: same idea with the formula here.

File: include/sampleflow/consumers/average_cosinus.h [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+ *
+ * ### Threading model ###
+ *
+ * The implementation of this class is thread-safe, i.e., its
+ * consume() member function can be called concurrently and from multiple
+ * threads.
+   **/
+
+    template <typename InputType>
+    class AverageCosinus: public Consumer<InputType>
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: The English word is actually "Cosine". I think you might want to change the name of the class. You might also want to think about naming it in a way that makes clear what it is the cosine of: Maybe `AverageCosineBetweenSuccessiveSamples`? That's a long name, but then there is always auto-completion.

File: include/sampleflow/consumers/average_cosinus.h [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+    {
+      public:
+        /**
+         * The data type of the elements of the input type.
+         */
+
+       using scalar_type = typename InputType::value_type;
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: no need for the empty line. Likewise below all of the other comments below -- they attach to the entity that follows, and there should be no empty line separating them.

Comment (bangerth) [reply include/sampleflow/consumers/average_cosinus.h]: By the way, what do you use this type for? Maybe that should also go into the documentation.

File: include/sampleflow/consumers/average_cosinus.h [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+
+       /**
+   	 * The data type, where we save multiple set of past samples.
+   	 */
+
+       using value_type = boost::numeric::ublas::matrix<scalar_type>;
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: The general convention is that the `value_type` is what is used as the return type of the `get()` function. If you need this type somewhere else, call it something different (and move the `using` declaration to the `private:` part of the class.

File: include/sampleflow/consumers/average_cosinus.h [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+
+        /**
+         * Constructor.
+         */
+
+        AverageCosinus(unsigned int length);
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]:
```suggestion
        AverageCosinus(const unsigned int length);
```

File: include/sampleflow/consumers/average_cosinus.h [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+
+        /**
+         * Process one sample by updating the previously computed covariance
+         * matrix using this one sample.
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: Might want to update the description here

File: include/sampleflow/consumers/average_cosinus.h [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+         */
+        virtual
+        void
+        consume (InputType     sample,
+                 AuxiliaryData aux_data) override;
+
+        /**
+         * A function that returns the average cosinus vector computed from the
+         * samples seen so far. If no samples have been processed so far, then
+         * a default-constructed object of type InputType will be returned.
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: The second sentence is not true.

File: include/sampleflow/consumers/average_cosinus.h [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+
+      mutable std::mutex mutex;
+
+
+        /**
+         * Parts for running average_cosinus
+         * Description of these parts is given above
+         */
+
+      std::vector<scalar_type> current_avg_cosinus;//Current average cosinus
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: no need to document this twice: what you have after the `//` should simply be in the `/** ... */` block

File: include/sampleflow/consumers/average_cosinus.h [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+
+        /**
+        * Save previous sample value
+        * %%%%%%%%%%%%Manty's comment: I didn't find (I'm 100% there are) better way to save last k sample values
+        * as just using one more object and assign i'th row of past_sample to i+1'th row of past_sample_replace
+        *
+        */
+
+      value_type past_sample;
+      value_type past_sample_replace;
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: I think you should be able to do `std::list<InputType>` and `push_back()` a new sample as it comes in, and `pop_front()` one if you no longer need it.

File: include/sampleflow/consumers/average_cosinus.h [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+
+        /**
+         * The number of samples processed so far.
+         */
+
+        types::sample_index n_samples;
+
+        unsigned int avg_cosinus_length;
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: I think you want to call this `history_length` or some such. I was confused by what the length here refers to.

File: include/sampleflow/consumers/average_cosinus.h [hunk 25]
```diff
@@ -0,0 +1,205 @@
...
+    };
+
+    template <typename InputType>
+    AverageCosinus<InputType>::
+	AverageCosinus (unsigned int length)
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]:
```suggestion
	AverageCosinus (const unsigned int length)
```

File: include/sampleflow/consumers/average_cosinus.h [hunk 26]
```diff
@@ -0,0 +1,205 @@
...
+          	  }
+
+          past_sample.resize(avg_cosinus_length,sample.size());
+          past_sample_replace.resize(avg_cosinus_length,sample.size());
+
+          for (unsigned int j=0; j<sample.size(); ++j) past_sample(0,j) = sample[j];
+        }
+
+      else
+		{
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: same here as elsewhere: Just indent the whole file.

File: include/sampleflow/consumers/average_cosinus.h [hunk 27]
```diff
@@ -0,0 +1,205 @@
...
+          n_samples = 1;
+
+          for (unsigned int i=0; i<avg_cosinus_length; ++i){
+        	  current_avg_cosinus[i] = 0;
+          	  }
+
+          past_sample.resize(avg_cosinus_length,sample.size());
+          past_sample_replace.resize(avg_cosinus_length,sample.size());
+
+          for (unsigned int j=0; j<sample.size(); ++j) past_sample(0,j) = sample[j];
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]:
```suggestion
          for (unsigned int j=0; j<sample.size(); ++j) 
            past_sample(0,j) = sample[j];
```

File: include/sampleflow/consumers/average_cosinus.h [hunk 28]
```diff
@@ -0,0 +1,205 @@
...
+    			}
+    	}
+
+    	for (unsigned int j=0; j<sample.size(); ++j) past_sample_replace(0,j) = sample[j];
+    	past_sample = past_sample_replace;
+
+	}
+   }
+
+//return value
```
Comment (bangerth) [inline include/sampleflow/consumers/average_cosinus.h]: just more comments you no longer need

File: include/sampleflow/consumers/covariance_matrix.h [hunk 29]
```diff
@@ -171,7 +171,7 @@ namespace SampleFlow
           delta -= current_mean;
           for (unsigned int i=0; i<sample.size(); ++i)
             for (unsigned int j=0; j<sample.size(); ++j)
-              current_covariance_matrix(i,j) += delta[i]*delta[j]/n_samples;
+              current_covariance_matrix(i,j) += delta[i]*delta[j]/n_samples-current_covariance_matrix(i,j)/(n_samples-1);
```
Comment (bangerth) [inline include/sampleflow/consumers/covariance_matrix.h]: This part of the patch is unrelated to everything else you do. It's often easiest to do this in a separate pull request.

Want me to extract this piece into something separate?

File: include/sampleflow/consumers/spurious_autocovariance_dim_n.h [hunk 30]
```diff
@@ -0,0 +1,289 @@
...
+
+#include <boost/numeric/ublas/matrix.hpp>
+
+
+namespace SampleFlow
+{
+namespace Consumers
+{
+/**
+ * NOTICE: We can only say, that this algorithm calculates "spurious autocovariance", because by definition
```
Comment (bangerth) [inline include/sampleflow/consumers/spurious_autocovariance_dim_n.h]: Make this a note at the bottom of the documentation, and use the syntax
```
  @note We can only say...
```

File: include/sampleflow/consumers/spurious_autocovariance_dim_n.h [hunk 31]
```diff
@@ -0,0 +1,289 @@
...
+ * autocovariance function is much more complex it is not (except if we work samples of one dimension). Anyway, later in this code we use word autocovariance as
+ * we refer to spurious autocovariance.
+ *
+ * This is a Consumer class that implements computing the running sample autocovariance function:
+ * $\hat\gamma(l)=\frac{1}{n}\sum_{t=1}^{n-l}{(\bm{x}_{t+l}-\bar\bm{x})(bm{x}_{t}-\bar\bm{x})}$
+ *
+ * This code for every new sample updates $\hat\gamma(k), l=1,2,3...,k.$ Choice of $autocovariance_length can be done
```
Comment (bangerth) [inline include/sampleflow/consumers/spurious_autocovariance_dim_n.h]:
```suggestion
 * This code for every new sample updates $\hat\gamma(k), l=1,2,3...,k$. The value of $n$ is set
```

File: include/sampleflow/consumers/spurious_autocovariance_dim_n.h [hunk 32]
```diff
@@ -0,0 +1,289 @@
...
+ * by setting it in mcmc_test.cc
```
Comment (bangerth) [inline include/sampleflow/consumers/spurious_autocovariance_dim_n.h]:
```suggestion
 * in the constructor.
```

File: include/sampleflow/consumers/spurious_autocovariance_dim_n.h [hunk 33]
```diff
@@ -0,0 +1,289 @@
...
+
+	using value_type = boost::numeric::ublas::matrix<scalar_type>;
+
+	/**
+	 * Constructor
+	 *
+	 * @param[in] parameter lag_lenght refers to how many autocovariance function values we want
+	 * to calculate
+	 */
+	SpuriousAutocovariance(unsigned int lag_length);
```
Comment (bangerth) [inline include/sampleflow/consumers/spurious_autocovariance_dim_n.h]:
```suggestion
	SpuriousAutocovariance(const unsigned int lag_length);
```

File: include/sampleflow/consumers/spurious_autocovariance_dim_n.h [hunk 34]
```diff
@@ -0,0 +1,289 @@
...
+	 *
+	 * @param[in] sample The sample to process.
+	 * @param[in] aux_data Auxiliary data about this sample. The current
+	 *   class does not know what to do with any such data and consequently
+	 *   simply ignores it.
+	 */
+	virtual
+	void
+	consume (InputType     sample,
+			AuxiliaryData aux_data) override;
```
Comment (bangerth) [inline include/sampleflow/consumers/spurious_autocovariance_dim_n.h]: Funny indentation -- just indent the whole file.

File: include/sampleflow/consumers/spurious_autocovariance_dim_n.h [hunk 35]
```diff
@@ -0,0 +1,289 @@
...
+	/**
+	 * A mutex used to lock access to all member variables when running
+	 * on multiple threads.
+	 */
+	mutable std::mutex mutex;
+
+	/**
+	 * Describes how many values of autocovariance function we calculate.
+	 */
+	unsigned int autocovariance_length;
```
Comment (bangerth) [inline include/sampleflow/consumers/spurious_autocovariance_dim_n.h]:
```suggestion
	const unsigned int autocovariance_length;
```

File: include/sampleflow/consumers/covariance_matrix.h (hunk 29 above)
Comment (MantautasRimkus) [reply include/sampleflow/consumers/covariance_matrix.h]: Yes, sure. Not sure how that "separate" looks like.

File: include/sampleflow/consumers/acceptance_ratio.h [hunk 36]
```diff
@@ -0,0 +1,166 @@
...
+          past_sample = sample;
+        }
+      //Check if new sample is not equal to previous sample and if that is true add one to accept.
+      else
+      	  {
+          ++n_samples;
+          bool truth=false;
+          for(int i = 0; i < sample.size(); i++) {
+              if(past_sample[i]!=sample[i]){
+            	  truth=true;
```
[the remark this answers is not in the document]
Comment (MantautasRimkus) [reply include/sampleflow/consumers/acceptance_ratio.h]: Okay. Then I'm leaving this untouched for this moment.

File: include/sampleflow/consumers/average_cosinus.h (hunk 23 above)
Comment (MantautasRimkus) [reply include/sampleflow/consumers/average_cosinus.h]: I use idea later, that I want to go through each element in a vector, which is a row of the matrix, like:
"norm2 += previous_sample[i][j]*previous_sample[i][j]; "

As I see at Google, there are no direct way how to achieve vectors in the list and elements of those vectors. I have googled a couple of different ways with pointers to do that (like https://thispointer.com/c-how-to-get-element-by-index-in-list/).
Do you have a suggestion which way would be preferred?

File: include/sampleflow/consumers/covariance_matrix.h (hunk 29 above)
Comment (bangerth) [reply include/sampleflow/consumers/covariance_matrix.h]: I can do that for you.

Comment (bangerth) [reply include/sampleflow/consumers/covariance_matrix.h]: Let's get everything else in order first, though.

File: include/sampleflow/consumers/average_cosinus.h (hunk 23 above)
Comment (bangerth) [reply include/sampleflow/consumers/average_cosinus.h]: Let's leave this for later as well.
